sway-ir/src/optimize/simplify_cfg.rs

```diff
@@ -33,43 +49,62 @@ pub fn simplify_cfg(context: &mut Context, function: &Function) -> Result<bool,
 }
 
 fn remove_dead_blocks(context: &mut Context, function: &Function) -> Result<bool, IrError> {
-    // Find a block other than 'entry' which has no predecessors, remove it if found.
-    function
-        .block_iter(context)
-        .skip(1)
-        .find(|block| block.num_predecessors(context) == 0)
-        .map(|dead_block| function.remove_block(context, &dead_block))
-        .transpose()
-        .map(|result| result.is_some())
-}
-
-fn merge_blocks(context: &mut Context, function: &Function) -> Result<bool, IrError> {
-    // Get the block a block branches to iff its terminator is `br`.
-    let block_terms_with_br = |from_block: Block| -> Option<(Block, Block)> {
-        from_block.get_terminator(context).and_then(|term| {
-            if let Instruction::Branch(to_block) = term {
-                Some((from_block, *to_block))
-            } else {
-                None
+    let mut worklist = std::collections::VecDeque::<Block>::new();
```

Comment:
I'm curious -- it might be slightly more efficient to just use a `Vec` here and `pop_back()`.  The order of visiting the blocks doesn't matter and while `VecDeque` is efficient it will need to do more allocations than `Vec` if you're going to use it as a queue.

Reply:
I was going to suggest you recurse instead of using a work list at all, but then the CFG for a large function could be quite deep so you'd need to guarantee tail calls to be sure not to blow up the stack.